fc/parsing/actions.py

```diff
@@ -1200,204 +1268,353 @@ def _number_generator(self, value, units):
-        else:
-            # Time units can also be set by an input specification, which will be handled correctly by the input
-            # handling code.
-            # If time units are set as an output, we need to handle it here: because time can be set from the protocol
-            # it is always an input, and so should follow the "input" rules for unit conversion.
-            for output in self.outputs:
-                try:
-                    variable = self.model.get_variable_by_ontology_term(output.rdf_term)
-                except KeyError:
-                    continue
+    def _collate_variable_information(self):
+        """Collates information from different variable defining clauses, and stores it in ProtocolVariable objects."""
 
-                if variable is self.time_variable and output.units is not None:
-                    units = self.units.get_unit(output.units)
-                    self.time_variable = self.model.convert_variable(variable, units, DataDirectionFlow.INPUT)
-                    break
+        # Temporary map from names to ProtocolVariable objects
+        name_to_pvar = {}
 
-    def _create_variable(self, var):
-        """Creates and returns an annotated variable based on the given :class:`VariableReference`."""
+        def get(ref):
+            try:
+                pvar = name_to_pvar[ref.prefixed_name]
+            except KeyError:
+                pvar = ProtocolVariable(ref.prefixed_name, ref.local_name, ref.rdf_term)
+                name_to_pvar[ref.prefixed_name] = pvar
 
-        name = self.model.get_unique_name('protocol__' + var.local_name)
-        units = self.units.get_unit(var.units)
-        variable = self.model.add_variable(name, units)
-        self.model.add_cmeta_id(variable)
-        self.model.rdf.add((variable.rdf_identity, PRED_IS, var.rdf_term))
-        return variable
+                # Store 'state_variable' pvar, if used in protocol
+                if ref.rdf_term == STATE_ANNOTATION:
+                    pvar.update(output_category=True)
+                    self.magic_pvar = pvar
 
-    def _add_input_variables(self):
-        """Ensure requested input variables exist, unless they are optional.
+            return pvar
 
-        If the variable exists its units are checked and converted if needed.
+        # Add inputs
+        #   input <prefix:term> [units <uname>] [= <initial_value>]
+        for ref in self.inputs:
+            pvar = get(ref)
+            pvar.update(input=True, units=ref.units, initial_value=ref.initial_value)
 
-        If it doesn't exist but is marked as optional, nothing is done.
+        # Add outputs and output categories
+        #   output <prefix:term> [units <uname>]
+        for ref in self.outputs:
+            pvar = get(ref)
+            pvar.update(output=True, units=ref.units)
+
+            # Check for "transitively" connected variables (e.g. the rdf term represents a category)
```

Comment:
OK, here we definitely have a problem, as the transitive search also (by design) picks up directly annotated variables, so you'll get length 1 vectors appearing here for every output variable.

I'm also very wary of treating the state vector differently from other vector outputs: doing so is likely to hide bugs with the general case. I _suspect_ (but haven't finished reviewing yet) doing the transitive search here for the general case means you're failing to update correctly after unit conversion, and this isn't picked up in the state vector case because it has magic support. In pycml the transitive lookup happens *after* all other manipulations.